clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp

```diff
@@ -354,6 +358,460 @@ fatbinary(ArrayRef<std::pair<StringRef, StringRef>> InputFiles,
+    return std::move(Err);
+  return *TempFileOrErr;
+}
+
+// This routine is used to convert SPIR-V input files into LLVM IR files.
+static Expected<StringRef> convertSPIRVToIR(StringRef Filename,
+                                            const ArgList &Args) {
+  Expected<std::string> SPIRVToIRWrapperPath = findProgram(
+      "spirv-to-ir-wrapper", {getMainExecutable("spirv-to-ir-wrapper")});
+  if (!SPIRVToIRWrapperPath)
+    return SPIRVToIRWrapperPath.takeError();
+
+  // Create a new file to write the converted file to.
+  auto TempFileOrErr =
+      createOutputFile(sys::path::filename(ExecutableName), "bc");
+  if (!TempFileOrErr)
+    return TempFileOrErr.takeError();
+
+  BumpPtrAllocator Alloc;
+  StringSaver Saver(Alloc);
+
+  SmallVector<StringRef, 8> CmdArgs;
+  CmdArgs.push_back(*SPIRVToIRWrapperPath);
+  CmdArgs.push_back(Filename);
+  CmdArgs.push_back("-o");
+  CmdArgs.push_back(*TempFileOrErr);
+  CmdArgs.push_back("-llvm-spirv-opts");
+  CmdArgs.push_back("\"--spirv-preserve-auxdata");
+  CmdArgs.push_back("--spirv-target-env=SPV-IR");
+  CmdArgs.push_back("--spirv-builtin-format=global\"");
+  if (Error Err = executeCommands(*SPIRVToIRWrapperPath, CmdArgs))
+    return std::move(Err);
+  return *TempFileOrErr;
+}
+
+// Run sycl-post-link tool
+static Expected<StringRef>
+runSYCLPostLink(SmallVectorImpl<StringRef> &InputFiles, const ArgList &Args) {
+  Expected<std::string> SYCLPostLinkPath =
+      findProgram("sycl-post-link", {getMainExecutable("sycl-post-link")});
+  if (!SYCLPostLinkPath)
+    return SYCLPostLinkPath.takeError();
+
+  // Create a new file to write the output of sycl-post-link to.
+  auto TempFileOrErr =
+      createOutputFile(sys::path::filename(ExecutableName), "table");
+  if (!TempFileOrErr)
+    return TempFileOrErr.takeError();
+
+  BumpPtrAllocator Alloc;
+  StringSaver Saver(Alloc);
+
+  StringRef SYCLPostLinkOptions;
+  if (Arg *A = Args.getLastArg(OPT_sycl_post_link_options_EQ))
+    SYCLPostLinkOptions = A->getValue();
+
+  SmallVector<StringRef, 8> CmdArgs;
+  CmdArgs.push_back(*SYCLPostLinkPath);
+  SYCLPostLinkOptions.split(CmdArgs, " ", /* MaxSplit = */ -1,
+                            /* KeepEmpty = */ false);
+  CmdArgs.push_back("-o");
+  CmdArgs.push_back(*TempFileOrErr);
+  CmdArgs.append(InputFiles);
+  if (Error Err = executeCommands(*SYCLPostLinkPath, CmdArgs))
+    return std::move(Err);
+  return *TempFileOrErr;
+}
+
+// This table is used to manage the output table populated by sycl-post-link.
+struct Table {
+  struct SYCLTableEntry {
+    std::string IRFile;
+    std::string PropFile;
+    std::string SymFile;
+  };
+
+  SmallVector<SYCLTableEntry, 16> Entries;
+
+  SmallVector<std::string, 16> getListOfIRFiles(void) {
+    SmallVector<std::string, 16> Files;
+    for (auto Entry : Entries) {
+      Files.push_back(Entry.IRFile);
+    }
+    return Files;
+  }
+
+  Expected<StringRef> writeSYCLTableToFile(void) {
```

Comment:
So right now we use the `file-table-tform` tool to manage all the table stuff. Are we trying to move again from that tool in the new flow? I saw the below stmt in the offloading doc, but we could move those tools to be shared libraries instead of executables and call it here, or move code into here or something but I'm not sure if we would reimplement them. 

For -fno-sycl-rdc mode, I had to enhance `file-table-tform` to support tables of tables and I'll need that here in the future, so it would be cool if we could reuse that code.

Text I was talking about:
Tools like llvm-foreach, file-table-tform, spirv-to-ir-wrapper were all introduced to provide a way to manipulate behaviors that could only be determined at runtime of the compiler toolchain. These were needed to work around the fact that the toolchain commands constructed by the driver is a fixed state of commands.

Reply:
The files are just a placeholder to hold tables. We will eventually be moving away from using files. We have a go opportunity her to remove such tools as we move from a 'static' compiler pipeline (in driver) to a dynamic one (in the linker-wrapper). Can you please add a link to the `file-table-tform` changes there and I can comment on how we can support in in the new flow.

Thanks

Reply:
The behavior I was talking about is [here](https://github.com/intel/llvm/blob/a01ae13fbe0144540fed43396a8a671be5dc1395/llvm/tools/file-table-tform/file-table-tform.cpp#L324), the merge functionality, thanks.

Reply:
AH..Ok. That should be easy to incorporate into the new flow.

Thanks